### ISSUE TITLE
`field_access_listener` is not able to link in VS2017

### ISSUE THREAD

**Issue by xhochy** [protocolbuffers/protobuf#8719]
**What version of protobuf and what language are you using?**
Version: v3.17.3 (v3.17.2 was still working, so this was introduced in v3.17.3)
Language: C++

**What operating system (Linux, Windows, ...) and version?**

Windows

**What runtime / compiler are you using (e.g., python version or gcc version)**

This is the conda-forge toolchain that basically boils down to using VS2017.

**What did you do?**

Build the C++ part of protobuf using CMake and VS2017 on Windows using the following build command:

```
cmake -G "Ninja" ^
         -DCMAKE_BUILD_TYPE=Release ^
         -DCMAKE_PREFIX_PATH=%LIBRARY_PREFIX% ^
         -DCMAKE_INSTALL_PREFIX=%LIBRARY_PREFIX% ^
         -Dprotobuf_WITH_ZLIB=ON ^
         -Dprotobuf_BUILD_SHARED_LIBS=ON ^
         -Dprotobuf_MSVC_STATIC_RUNTIME=OFF ^
         ..
```

We get the following error: 

```
2021-06-08T18:31:58.5615690Z    Creating library libprotobuf-lite.lib and object libprotobuf-lite.exp
2021-06-08T18:31:58.5618288Z field_access_listener.cc.obj : error LNK2019: unresolved external symbol "private: void __cdecl google::protobuf::internal::LazyDescriptor::Once(void)" (?Once@LazyDescriptor@internal@protobuf@google@@AEAAXXZ) referenced in function "public: class google::protobuf::Descriptor const * __cdecl google::protobuf::internal::LazyDescriptor::Get(void)" (?Get@LazyDescriptor@internal@protobuf@google@@QEAAPEBVDescriptor@34@XZ)
2021-06-08T18:31:58.5623225Z field_access_listener.cc.obj : error LNK2019: unresolved external symbol "public: struct google::protobuf::Descriptor::ExtensionRange const * __cdecl google::protobuf::Descriptor::FindExtensionRangeContainingNumber(int)const " (?FindExtensionRangeContainingNumber@Descriptor@protobuf@google@@QEBAPEBUExtensionRange@123@H@Z) referenced in function "public: bool __cdecl google::protobuf::Descriptor::IsExtensionNumber(int)const " (?IsExtensionNumber@Descriptor@protobuf@google@@QEBA_NH@Z)
2021-06-08T18:31:58.5630246Z field_access_listener.cc.obj : error LNK2019: unresolved external symbol "public: struct google::protobuf::Descriptor::ReservedRange const * __cdecl google::protobuf::Descriptor::FindReservedRangeContainingNumber(int)const " (?FindReservedRangeContainingNumber@Descriptor@protobuf@google@@QEBAPEBUReservedRange@123@H@Z) referenced in function "public: bool __cdecl google::protobuf::Descriptor::IsReservedNumber(int)const " (?IsReservedNumber@Descriptor@protobuf@google@@QEBA_NH@Z)
2021-06-08T18:31:58.5637942Z field_access_listener.cc.obj : error LNK2019: unresolved external symbol "private: bool __cdecl google::protobuf::FieldDescriptor::is_map_message_type(void)const " (?is_map_message_type@FieldDescriptor@protobuf@google@@AEBA_NXZ) referenced in function "public: bool __cdecl google::protobuf::FieldDescriptor::is_map(void)const " (?is_map@FieldDescriptor@protobuf@google@@QEBA_NXZ)
2021-06-08T18:31:58.5642641Z field_access_listener.cc.obj : error LNK2019: unresolved external symbol "private: static void __cdecl google::protobuf::FieldDescriptor::TypeOnceInit(class google::protobuf::FieldDescriptor const *)" (?TypeOnceInit@FieldDescriptor@protobuf@google@@CAXPEBV123@@Z) referenced in function "public: enum google::protobuf::FieldDescriptor::Type __cdecl google::protobuf::FieldDescriptor::type(void)const " (?type@FieldDescriptor@protobuf@google@@QEBA?AW4Type@123@XZ)
2021-06-08T18:31:58.5646599Z field_access_listener.cc.obj : error LNK2019: unresolved external symbol "public: struct google::protobuf::EnumDescriptor::ReservedRange const * __cdecl google::protobuf::EnumDescriptor::FindReservedRangeContainingNumber(int)const " (?FindReservedRangeContainingNumber@EnumDescriptor@protobuf@google@@QEBAPEBUReservedRange@123@H@Z) referenced in function "public: bool __cdecl google::protobuf::EnumDescriptor::IsReservedNumber(int)const " (?IsReservedNumber@EnumDescriptor@protobuf@google@@QEBA_NH@Z)
2021-06-08T18:31:58.5651471Z field_access_listener.cc.obj : error LNK2019: unresolved external symbol "public: class google::protobuf::FileDescriptor const * __cdecl google::protobuf::FileDescriptor::dependency(int)const " (?dependency@FileDescriptor@protobuf@google@@QEBAPEBV123@H@Z) referenced in function "public: class google::protobuf::FileDescriptor const * __cdecl google::protobuf::FileDescriptor::public_dependency(int)const " (?public_dependency@FileDescriptor@protobuf@google@@QEBAPEBV123@H@Z)
2021-06-08T18:31:58.5655336Z field_access_listener.cc.obj : error LNK2019: unresolved external symbol "public: virtual __cdecl google::protobuf::DescriptorPool::ErrorCollector::~ErrorCollector(void)" (??1ErrorCollector@DescriptorPool@protobuf@google@@UEAA@XZ) referenced in function "public: virtual void * __cdecl google::protobuf::DescriptorPool::ErrorCollector::`vector deleting destructor'(unsigned int)" (??_EErrorCollector@DescriptorPool@protobuf@google@@UEAAPEAXI@Z)
2021-06-08T18:31:58.5658822Z field_access_listener.cc.obj : error LNK2019: unresolved external symbol "public: void __cdecl google::protobuf::DescriptorPool::InternalDontEnforceDependencies(void)" (?InternalDontEnforceDependencies@DescriptorPool@protobuf@google@@QEAAXXZ) referenced in function "public: void __cdecl google::protobuf::DescriptorPool::InternalSetLazilyBuildDependencies(void)" (?InternalSetLazilyBuildDependencies@DescriptorPool@protobuf@google@@QEAAXXZ)
2021-06-08T18:31:58.5728661Z field_access_listener.cc.obj : error LNK2019: unresolved external symbol "private: static enum google::protobuf::FieldDescriptor::CppType const * const google::protobuf::FieldDescriptor::kTypeToCppTypeMap" (?kTypeToCppTypeMap@FieldDescriptor@protobuf@google@@0QBW4CppType@123@B) referenced in function "public: static enum google::protobuf::FieldDescriptor::CppType __cdecl google::protobuf::FieldDescriptor::TypeToCppType(enum google::protobuf::FieldDescriptor::Type)" (?TypeToCppType@FieldDescriptor@protobuf@google@@SA?AW4CppType@123@W4Type@123@@Z)
2021-06-08T18:31:58.5732944Z field_access_listener.cc.obj : error LNK2019: unresolved external symbol "private: static char const * const * const google::protobuf::FieldDescriptor::kTypeToName" (?kTypeToName@FieldDescriptor@protobuf@google@@0QBQEBDB) referenced in function "public: static char const * __cdecl google::protobuf::FieldDescriptor::TypeName(enum google::protobuf::FieldDescriptor::Type)" (?TypeName@FieldDescriptor@protobuf@google@@SAPEBDW4Type@123@@Z)
2021-06-08T18:31:58.5736422Z field_access_listener.cc.obj : error LNK2019: unresolved external symbol "private: static char const * const * const google::protobuf::FieldDescriptor::kCppTypeToName" (?kCppTypeToName@FieldDescriptor@protobuf@google@@0QBQEBDB) referenced in function "public: static char const * __cdecl google::protobuf::FieldDescriptor::CppTypeName(enum google::protobuf::FieldDescriptor::CppType)" (?CppTypeName@FieldDescriptor@protobuf@google@@SAPEBDW4CppType@123@@Z)
2021-06-08T18:31:58.5739403Z libprotobuf-lite.dll : fatal error LNK1120: 12 unresolved externals
```

Full build log: 
[build.log](https://github.com/protocolbuffers/protobuf/files/6626290/build.log)

**Comment by OgreTransporter** [comment]
I have the same problem with VS2019.

**Comment by OgreTransporter** [comment]
https://github.com/protocolbuffers/protobuf/blob/e7440d2f13565815e66a200aea05ae4ce8f988e6/src/google/protobuf/field_access_listener.h#L104

FieldDescriptor is not part of libprotobuf-lite (field_access_listener).

**Comment by a2824256** [comment]
I have the same problem with VS2019，too.

**Comment by a2824256** [comment]
@xhochy @OgreTransporter Hey guys, I have solved this problem. 
Firstly, i clone the source code, then use the following code to download the submodule.
```
git submodule update --init --recursive
```
The problem with the above statement may cause the downloaded submodule to not correspond to the protobuf version.

Finally, I downloaded the compressed package of release and successfully compiled protobuf.
The [release zip](https://github.com/protocolbuffers/protobuf/releases)

The compile environment is win10 + vs2019 + cmake 3.21
